### PR TITLE
Fixing bug that even leads incompatible string to be passed to Integer.parseInt

### PR DESCRIPTION
**Pull request by Shilpa-Rani**
Initially, some of the tests failed, because the following logic in SimpleMMcifParser.java does not prevent invalid strings from being passed to Integer.parseInt()

`if ( val != null && ! val.equals("?") && !val.equals("."))`
If val is either "?" or ".", it will still return a true because of the presence of && operator. 

I have written a generic function that would prevent any invalid string from being passed to an Integer.parseInt():
The module checkIfValidStrForIntConversion_ returns a boolean value. The logic states that if the string to be parsed to Integer is either null or empty or contains any characters apart from digits 0-9, the method would return a false. In all other cases, it will return a true.

### CODE REVIEW

File: biojava-structure/src/main/java/org/biojava/nbio/structure/io/mmcif/SimpleMMcifParser.java [hunk 1]
```diff
@@ -1021,7 +1021,8 @@ private Object buildObject(String className, List<String> loopFields, List<Strin
 
 			try {
 				if ( pType[0].getName().equals(Integer.class.getName())) {
-					if ( val != null && ! val.equals("?") && !val.equals(".")) {
+// 					if ( val != null && ! val.equals("?") && !val.equals(".")) {
```
Comment (sbliven) [inline biojava-structure/src/main/java/org/biojava/nbio/structure/io/mmcif/SimpleMMcifParser.java:1024]: Please remove the comment.

**Review by sbliven — CHANGES_REQUESTED** [review]
I'm confused by this issue. Can you provide an example file that doesn't parse?

Contrary to your comment, the existing versions works fine for "?" and "." (`!` is higher than `&&` in the order of operations). Your version does exclude arbitrary strings ("x") but it fails for negative numbers and other edge cases. Furthermore, if the file reaches this code than it is expecting a field of type Integer, so throwing an exception for invalid input seems acceptable.

File: biojava-structure/src/main/java/org/biojava/nbio/structure/io/mmcif/SimpleMMcifParser.java (hunk 1 above)
Comment (Vidishab18) [reply biojava-structure/src/main/java/org/biojava/nbio/structure/io/mmcif/SimpleMMcifParser.java:1024]: We acknowledge that the current module doesn’t handle negative numbers and we would soon provide the modified implementation for the same. 

The example scenarios that leads to parsing errors are below:

```
[ERROR] Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 2.827 s <<< FAILURE! - in org.biojava.nbio.structure.io.TestMMCIFWriting
[ERROR] test1A2C(org.biojava.nbio.structure.io.TestMMCIFWriting)  Time elapsed: 2.826 s  <<< ERROR!
java.lang.NumberFormatException: For input string: "?"
	at org.biojava.nbio.structure.io.TestMMCIFWriting.testRoundTrip(TestMMCIFWriting.java:153)
	at org.biojava.nbio.structure.io.TestMMCIFWriting.test1A2C(TestMMCIFWriting.java:77)

[ERROR] Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 2.843 s <<< FAILURE! - in org.biojava.nbio.structure.io.TestMMCIFWriting
[ERROR] test1A2C(org.biojava.nbio.structure.io.TestMMCIFWriting)  Time elapsed: 2.838 s  <<< ERROR!
java.lang.NumberFormatException: For input string: "A"
	at org.biojava.nbio.structure.io.TestMMCIFWriting.testRoundTrip(TestMMCIFWriting.java:153)
	at org.biojava.nbio.structure.io.TestMMCIFWriting.test1A2C(TestMMCIFWriting.java:77)

[ERROR] test1A2C(org.biojava.nbio.structure.io.TestMMCIFWriting)  Time elapsed: 2.816 s  <<< ERROR!
java.lang.NumberFormatException: For input string: "?"
	at org.biojava.nbio.structure.io.TestMMCIFWriting.testRoundTrip(TestMMCIFWriting.java:153)
	at org.biojava.nbio.structure.io.TestMMCIFWriting.test1A2C(TestMMCIFWriting.java:77)

[ERROR] test1A2C(org.biojava.nbio.structure.io.TestMMCIFWriting)  Time elapsed: 2.825 s  <<< ERROR!
java.lang.NumberFormatException: For input string: "N"
	at org.biojava.nbio.structure.io.TestMMCIFWriting.testRoundTrip(TestMMCIFWriting.java:153)
	at org.biojava.nbio.structure.io.TestMMCIFWriting.test1A2C(TestMMCIFWriting.java:77)

[ERROR] Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 2.711 s <<< FAILURE! - in org.biojava.nbio.structure.io.TestMMCIFWriting
[ERROR] test1A2C(org.biojava.nbio.structure.io.TestMMCIFWriting)  Time elapsed: 2.703 s  <<< ERROR!
java.lang.NumberFormatException: For input string: "?"
	at org.biojava.nbio.structure.io.TestMMCIFWriting.testRoundTrip(TestMMCIFWriting.java:153)
	at org.biojava.nbio.structure.io.TestMMCIFWriting.test1A2C(TestMMCIFWriting.java:77)

[ERROR] Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 2.727 s <<< FAILURE! - in org.biojava.nbio.structure.io.TestMMCIFWriting
[ERROR] test1A2C(org.biojava.nbio.structure.io.TestMMCIFWriting)  Time elapsed: 2.723 s  <<< ERROR!
java.lang.NumberFormatException: For input string: "THR"
	at org.biojava.nbio.structure.io.TestMMCIFWriting.testRoundTrip(TestMMCIFWriting.java:153)
	at org.biojava.nbio.structure.io.TestMMCIFWriting.test1A2C(TestMMCIFWriting.java:77)

[ERROR] test1A2C(org.biojava.nbio.structure.io.TestMMCIFWriting)  Time elapsed: 2.655 s  <<< ERROR!
java.lang.NumberFormatException: For input string: "50.00"
	at org.biojava.nbio.structure.io.TestMMCIFWriting.testRoundTrip(TestMMCIFWriting.java:153)
	at org.biojava.nbio.structure.io.TestMMCIFWriting.test1A2C(TestMMCIFWriting.java:77)

[ERROR] test1A2C(org.biojava.nbio.structure.io.TestMMCIFWriting)  Time elapsed: 2.814 s  <<< ERROR!
java.lang.NumberFormatException: For input string: "12.307"
	at org.biojava.nbio.structure.io.TestMMCIFWriting.testRoundTrip(TestMMCIFWriting.java:153)
	at org.biojava.nbio.structure.io.TestMMCIFWriting.test1A2C(TestMMCIFWriting.java:77)

[ERROR] test1A2C(org.biojava.nbio.structure.io.TestMMCIFWriting)  Time elapsed: 2.723 s  <<< ERROR!
java.lang.NumberFormatException: For input string: "1.00"
	at org.biojava.nbio.structure.io.TestMMCIFWriting.testRoundTrip(TestMMCIFWriting.java:153)
	at org.biojava.nbio.structure.io.TestMMCIFWriting.test1A2C(TestMMCIFWriting.java:77)

[ERROR] Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 2.177 s <<< FAILURE! - in org.biojava.nbio.structure.io.TestMMCIFWriting
[ERROR] test1SMT(org.biojava.nbio.structure.io.TestMMCIFWriting)  Time elapsed: 2.174 s  <<< ERROR!
java.lang.NumberFormatException: For input string: "-6.786"
	at org.biojava.nbio.structure.io.TestMMCIFWriting.testRoundTrip(TestMMCIFWriting.java:153)
	at org.biojava.nbio.structure.io.TestMMCIFWriting.test1SMT(TestMMCIFWriting.java:60) 
```

The logs are retrieved from running the Nondex tool on biojava-structure module.

**Comment by josemduarte** [comment]
> if the file reaches this code than it is expecting a field of type Integer, so throwing an exception for invalid input seems acceptable.

I agree with @sbliven 's comment. 

Additionally it is not very clear that this is the real issue in `TestMMCIFWriting` test that you provide as an example. I am guessing there is probably a different underlying issue.

**Pull request closed without merging**